[cert-manager] invalid peer certificate: UnknownIssuer

I have a cluster with mtls set up as shown below. I originally had the certificate durations much shorter but updated to the length below.

```yaml
  mTLS:
    certificatesDuration:
      clientCertificates: 720h0m0s
      frontendCertificate: 24h0m0s
      intermediateCAsCertificates: 1440h0m0s
      internodeCertificate: 24h0m0s
      rootCACertificate: 2160h0m0s
    frontend:
      enabled: true
    internode:
      enabled: true
    provider: cert-manager
    refreshInterval: 1h0m0s
```

Later I attempted to connect via a client using a certificate from the secret created via `TemporalClusterClient` and got `invalid peer certificate: UnknownIssuer`. I then:
* deleted and recreated the `TemporalClusterClient` and used the new secret and still had same issue.
* deleted the front end deployment for temporal and waited for it to be recreated and still had same issue
* deleted the frontend intermediate CA secret, the frontend certificate secret and the frontend deployment and then the connection worked.

I'm not sure if this was due to me updating the certificate durations on `TemporalCluster` or some other issue but somehow the certificates were not properly renewed. I'll keep experimenting with this but opened the issue in case others have similar problems.

> @alexandrevilain I tested with the 0.16.1 operator version and observed that the secret created by temporalclusterclient is not synced in the worker namespace after the certificate is renewed. Both the secrets i.e. in temporal cluster namespace and in worker namespace differ in content, and we observe bad/invalid certificate errors. Could you please check?
> 
> Below is the manifest that I used -
> 
> ```
> apiVersion: temporal.io/v1beta1
> kind: TemporalClusterClient
> metadata:
>   name: my-worker
>   namespace: my-worker-ns
> spec:
>   clusterRef:
>     name: temporal-cluster-dev
>     namespace: temporal-cluster-ns
> ```
> 
> Below is the mtls configuration in temporal cluster manifest. We have put small duration to test the secret sync on certificate expiration and renewal.
> 
> ```
> mTLS:
>   certificatesDuration:
>     clientCertificates: 1h30m
>     frontendCertificate: 1h30m
>     intermediateCAsCertificates: 1h30m
>     internodeCertificate: 1h30m
>     rootCACertificate: 17520h
>   frontend:
>     enabled: true
>   internode:
>     enabled: true
>   refreshInterval: 30m
>   renewBefore: 5m
> ```
> 
> Our worker logs once the certs get expired (and secret is not updated on renewal) -
> 
> ```
> Caused by: javax.net.ssl.SSLException: error:10000412:SSL routines:OPENSSL_internal:SSLV3_ALERT_BAD_CERTIFICATE
> 	at io.grpc.netty.shaded.io.netty.handler.ssl.ReferenceCountedOpenSslEngine.shutdownWithError(ReferenceCountedOpenSslEngine.java:1083)
> 	at io.grpc.netty.shaded.io.netty.handler.ssl.ReferenceCountedOpenSslEngine.sslReadErrorResult(ReferenceCountedOpenSslEngine.java:1377)
> 	at io.grpc.netty.shaded.io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1317)
> 	at io.grpc.netty.shaded.io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1404)
> 	at io.grpc.netty.shaded.io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1447)
> 	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler$SslEngineType$1.unwrap(SslHandler.java:222)
> 	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1343)
> 	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1236)
> 	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1285)
> 	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
> 	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)
> 	... 15 common frames omitted
> ```
> 
> Also, I don't see any errors in the temporal-operator logs.


@alexandrevilain Could you please take a look at the issue? Is there anything that I am missing? Thanks.

--- a/controllers/temporalclusterclient_controller.go
+++ b/controllers/temporalclusterclient_controller.go
@@ -28,10 +28,12 @@ import (
 	certmanagermeta "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
 	kerrors "k8s.io/apimachinery/pkg/util/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
@@ -47,6 +49,11 @@ type TemporalClusterClientReconciler struct {
 
 	AvailableAPIs *discovery.AvailableAPIs
 }
+
+var (
+	clusterRefNameField      = "spec.clusterRef.name"
+	clusterRefNamespaceField = "spec.clusterRef.namespace"
+)
 
 //+kubebuilder:rbac:groups=temporal.io,resources=temporalclusterclients,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=temporal.io,resources=temporalclusterclients/status,verbs=get;update;patch
@@ -148,13 +155,87 @@ func (r *TemporalClusterClientReconciler) SetupWithManager(mgr ctrl.Manager) err
 	controller := ctrl.NewControllerManagedBy(mgr).
 		For(&v1beta1.TemporalClusterClient{})
 
+	err := mgr.GetFieldIndexer().IndexField(
+		context.Background(),
+		&v1beta1.TemporalClusterClient{},
+		clusterRefNameField,
+		func(rawObj client.Object) []string {
+			clusterClient := rawObj.(*v1beta1.TemporalClusterClient)
+			return []string{clusterClient.Spec.ClusterRef.Name}
+		})
+	if err != nil {
+		return err
+	}
+
+	err = mgr.GetFieldIndexer().IndexField(
+		context.Background(),
+		&v1beta1.TemporalClusterClient{},
+		clusterRefNamespaceField,
+		func(rawObj client.Object) []string {
+			clusterClient := rawObj.(*v1beta1.TemporalClusterClient)
+			return []string{clusterClient.Spec.ClusterRef.Namespace}
+		})
+	if err != nil {
+		return err
+	}
+
 	if r.AvailableAPIs.CertManager {
 		controller = controller.
-			Owns(&certmanagerv1.Issuer{}).
-			Owns(&certmanagerv1.Certificate{})
+			Watches(
+				&certmanagerv1.Certificate{},
+				handler.EnqueueRequestsFromMapFunc(
+					EnqueueRequestForClusterClientReferencingOwnerCluster(r.Client),
+				)).
+			Watches(
+				&certmanagerv1.Issuer{},
+				handler.EnqueueRequestsFromMapFunc(
+					EnqueueRequestForClusterClientReferencingOwnerCluster(r.Client),
+				))
 	}
 
 	controller.Owns(&corev1.Secret{})
 
 	return controller.Complete(r)
+}
+
+// EnqueueRequestForClusterClientReferencingOwnerCluster returns a reconcile request for any TemporalClusterClient
+// referencing the given the TemporalCluster owner for the provided object.
+func EnqueueRequestForClusterClientReferencingOwnerCluster(c client.Client) handler.MapFunc {
+	return func(ctx context.Context, object client.Object) []reconcile.Request {
+		log := log.FromContext(ctx)
+		result := []reconcile.Request{}
+
+		// from the object get the owner cluster.
+		// then get the cluster client from the cluster.
+		for _, owner := range object.GetOwnerReferences() {
+			if owner.Kind != v1beta1.TemporalClusterTypeMeta.Kind {
+				continue
+			}
+			list := &v1beta1.TemporalClusterClientList{}
+			err := c.List(ctx,
+				list,
+				client.MatchingFields{
+					clusterRefNameField: owner.Name,
+				},
+				client.MatchingFields{
+					clusterRefNamespaceField: object.GetNamespace(),
+				},
+			)
+			if err != nil {
+				log.Error(err, "Failed to get TemporalClusterClient referencing owner TemporalCluster, skipping mapping.")
+				return nil
+			}
+
+			for _, client := range list.Items {
+				result = append(result, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      client.GetName(),
+						Namespace: client.GetNamespace(),
+					},
+				})
+			}
+		}
+
+		return result
+	}
 }

